README.md
# The Pahoehoe Encrypted Proxy Project
The client and server code used by Lavabit to operate its encrypted proxy service. This repo contains the code for Android client application. It also contains the scripts we use to automate the building, and testing of that app.

These same scripts can also be used to setup local VPN instances for testing, or be adapted to setup production VPN servers for those interested in forking the app, and creating their own variation.

[<img src="https://fdroid.gitlab.io/artwork/badge/get-it-on.png"
     alt="Get it on F-Droid"
     height="80">](https://f-droid.org/packages/com.lavabit.pahoehoe/)
[<img src="https://play.google.com/intl/en_us/badges/images/generic/en-play-badge.png"
     alt="Get it on Google Play"
     height="80">](https://play.google.com/store/apps/details?id=com.lavabit.pahoehoe)

### Building (Easy Method)
To build the code in this repo, clone the repo and execute the `run.sh` script. That script will setup Alma and Debian virtual machines as local VPN servers, which you can then test against, and it will setup a third Debian virtual machine with the dependencies needed to build the Android APK files. If you would like to build the APK without using the Vagrant VMs consult the `debian-10-build-setup.sh` to see what dependencies are required, and the `debian-10-build.sh` for how to build the APK using the command line.

Note that building the APK using Android Studio will fail unless you manually build the VPN libraries using the command line, before importing the project. Use the `android/scripts/build_deps.sh` to build the binary libraries via the command line.

**Release Checklist**   
To create a new release, update the `run.sh` and the `android/app/build.gradle` files. Remove the `RC` from the version string, and commit the code base. Then create a matching `git tag`. While finalizing a release build and generating the files for the Play Store, and website, it's critical that the `VERSTR` and `VERCODE` values in the `run.sh` script match up with the values in `android/app/build.gradle`, at least those values used by the Lavabit variant. And then it's also critical that the procelain commit gets tagged with an identical version string. This will ensure the F-Droid store is updated, and that their version strings match those distributed elsewhere.

Once the release files have been generated, increment the version code and version string values used by the `run.sh` and `android/app/build.gradle` files, while also appending `RC` back onto the end of the version string. Please also note that the `android-11-installer.sh` and `debian-10-emulator.sh` should be updated with the new version string values (including the RC), so they can find properly predict the build filenames to install.

**Direct Access**
For those looking to use the Lavabit proxy servers directly, and without the assistance of our apps, the `lavabitvpn.sh` script may provide a guide. It tries to create a proxy client configuration that can be used directly with any generic OpenVPN client. Unfortunately this script is rather fragile, and may require root permissions on some systems, so it can properly safeguard the configuration it generates. We've also found that even with root, it may not work properly on all distros.

### Tasks
  
- Update the config so the automatic retry window will only grow to a maxium of 10 seconds.
- Randomize which proxy node is used if multiple  nodes are online at a single location.
- Remove the QUERY_ALL_PACKAGES permission from the manifest, or fix and restore access to the excluder.
- Allow users to manually switch to using UDP, instead of TCP.
- Properly implement dnsmasq on the proxy nodes, so that DNS queries get tunneled properly.
- Once the DNS issue is fixed, submit a pull request upstream with our certificate and requisite config info.
- Add a caching layer to the VPNweb daemon to reduce the lag a user might face during setup.
- Adapt the updated description we use on the Play Store for F-droid, and enable the translations.
- Update the OpenVPN version on the server, and along with the embedded OpenVPN client, and enable TLS 1.3.
- Incorporate post quantum cipher support (and allow users to enable their use).
- Update the SDK target from 30 (Android 11) to 32 (Android 12L), or even 33 (Android 13).
- Begin uploading our Android app to the Samsung and Amazon app stores, and allow direct downloads via the Lavabit website.
- Potentially offer the app via APKPure, Uptodow, SlideMe and Aptoide.
  
<sub><sub>The Aurora/Yalp stores are Play Store proxies, and the Huawei, Xiaomi, OPPO and VIVO stores are based in China, where VPN apps are illegal.</sub></sub>
  
**Future Release Work**
Integrate support for the [Google Play Publisher](https://github.com/Triple-T/gradle-play-publisher/blob/master/README.md#managing-play-store-metadata) Gradle radleplugin which will allow us to automate the submission of new builds and the management of metadata in the same way we already integrate with the F-Droid store.

### Security Note
The production releases keys/creation is handled by the `debian-10-build-key.sh` script module. Since this script is likely to contain sensitive information, like the release signing keys, it should never be committed to the repo, or pushed to a server. The git ignore directive should prevent this, but for extra security, consider adding the following script as `.git/hooks/pre-commit` to the production build system git repo:

```
#! /bin/sh -e
git ls-files --cached | grep -qx 'debian-10-build-key.sh' && { echo "The release build keys file has been included in this commit. Remove it and try again." >&2; exit 1; }
exit 0
```

### Upstream Code
While the management, and build scripts for the project were developed by Lavabit, and are unique to this repo, the client and server code wasn't created from nothingness. Rather it is the byproduct of a polygamus relationship between several pre-existing F/OSS projects.

Our Android application code is a fork of the [Bitmask codebase](https://0xacab.org/leap/bitmask_android), which acts as a wrapper around the [ICS OpenVPN project](https://github.com/schwabe/ics-openvpn).

The HTTPS server daemon we use to facilitate client setup, and provide access credentials is an adaptation of the [VPNweb repository](https://0xacab.org/leap/vpnweb/). The server utilizes a several Go modules, all of which are also F/OSS, to function.

All of the code developed by Lavabit is hereby released under the GPL. Most of the code we incorporate, but not all of it, is also available under the GPL. Just beware that some of the code you may find buried in sub-folder may be subject to a different license. Even though the license may not be the GPL, all of the licenses are OSI approved.


daemon/CHANGELOG
CHANGELOG
=========

0.8 (20210304)

    - Fixed formatting typo in README.md
    - Added support for (Linux only) musl libc by including <sys/ttydefaults.h> for CEOF (by Baruch Siach)
    - Added --bind option to terminate when the user logs out (iff systemd/elogind) (with ZhaoLin1457 et al.)
    - Added Slackware-specific configuration
    - Added expansion of $VAR and ~user notation in command line option arguments and configuration files
    - Added environment variable definitions in configuration files
    - Automatically create --pidfiles or --pidfile directory if it's under the user's home directory
    - Updated README.md (improved notes on --stop and --restart, and added --signal --running --list --bind)
    - Added configure --help, --destdir, --enable-logind and --disable-logind options
    - Fixed $(DESTDIR) use in makefiles to interoperate with configure --prefix

0.7.1 (20210220)

    - Fixed make distclean/clobber on Debian (use debian-clobber instead of clobber) (spotted by Marc Haber)
    - Fixed parallel make test (race condition creating test directory) (spotted by Marc Haber)
    - Fixed manpage header and footer and various other minor fixes and improvements
    - Minor improvements in a few module tests
    - Added configure --prefix=/opt/local for macports (spotted by Kurt Hindenburg)
    - Added mention of file locations on macOS/macports to the daemon(1) manpage
    - Added checks for gcc -W options that fail on old macosx (e.g. 10.6.8) without macports
    - Minor changes to usage error messages for consistency
    - Improved config file parsing (skip spaces around commas and added more error messages)
    - Added support for using $CC in configure (spotted by Ryan Schmidt)
    - Fixed Makefile (wasn't inheriting CPPFLAGS, CFLAGS or LDFLAGS from caller) (spotted by Ryan Schmidt)

0.7 (20201111)

    - Fixed bug: race condition causing failure with "exec daemon ..." (with Askar Safin safinaskar at mail.ru)
    - Added $DAEMON_INIT_EXIT_DELAY_MSEC workaround for obscure KDE/konsole bug (with Askar Safin safinaskar at mail.ru)
    - Removed O_NOCTTY-related manpage "bug" (now historical) (spotted by Askar Safin safinaskar at mail.ru)
    - Always double fork for safety on systems of unknown flavour (suggested by Askar Safin safinaskar at mail.ru)
    - Added support for GNU/Hurd (by Svante Signell svante.signell at telia.com)
    - Added --ignore-eof and --read-eof options to alter behaviour when client dies without closing stdio
      (for Terry Greeniaus greeniaus_terry at emc.com and Tilman Baumann at tilman.baumann at grandeye.com)
    - Fixed bug: daemon_lock_pidfile() leaked fd across exec (spotted by Tilman Baumann tilman.baumann at grandeye.com)
    - Added --signal option (suggested by Marc Haber mh+daemon-dev at zugschlus.de)
    - Added --nocore option and removed silly manpage "bug" about its absence
    - Added fcntl_set_fdflag() and fcntl_clear_fdflag() to the fio module
    - Added signal_set_siginfo_handler() to the sig module
    - Added msg_filter_t, msg_create_filter(), msg_create_filter_with_locker() to the msg module
    - Test secure memory in the mem module for non-root users as well (now that they can do it)
    - Added prog_{out,err,dbg,alert}_push_filter() to the prog module
    - Added recvcred(), recvfromcred() to the net module
    - Added a few more debug messages that were missing
    - Insert daemon name before client's argv[0] so it appears in ps output (for marko.asplund at gmail.com)
    - Allow --pidfiles to be present without --name also being present
    - Added --list option to print a list of named daemons (more detail with --verbose)
    - Improved error message when client exit()s after --stop (for jakubholy.net at gmail.com)
    - Added syslog facility and priority lists to daemon(1) manpage (for jakubholy.net at gmail.com)
    - Various fixes and improvements to manpage
    - Improved --verbose --running message to state if client is running as well or not (added client pid file)
    - Fixed buglet: daemon no longer exits with EXIT_FAILURE when client is killed by an unexpected signal
    - Fixed harmless format string bug in prog_opt_process() (spotted by Jong-Gwon Kim and Woosuk Lee {jgkim,wslee} at ropas.snu.ac.kr)
    - Fixed bug: leaked fds for --errlog, --dbglog, --stdout and --stderr across exec
    - Fixed bug in parallelized make: "make ready" first so do it in configure (spotted by ilovezfs at icloud.com)
    - Fixed copyright messages (list actual years of publication, not just a range of years)
    - Fixed all sscanf "%s" to limit output (e.g. "%63s") to prevent buffer overruns (was only in test and example code)
    - Added more tests in the prop module
    - Fixed buglet: Removed vhangup() in pty_make_controlling_tty() as it broke some coproc module tests if run as root
    - Fixed str.h to not clash with new strlcpy/strlcat macros on OSX/macOS
    - Updated to avoid new warnings in recent versions of gcc
    - Split README into multiple files (README.md INSTALL COPYING REFERENCES CHANGELOG)
    - Added /etc/daemon.conf.d/ and ~/.daemonrc.d/ configuration directories
    - Changed "config" script name to "configure" (not an alias anymore) (but it's still not GNU configure)
    - Changed /etc/daemon.conf file location to /usr/local/etc/daemon.conf on FreeBSD
    - Changed /etc/daemon.conf file location to /usr/pkg/etc/daemon.conf on NetBSD
    - Changed installation location prefix to /usr/pkg on NetBSD
    - Added concrete examples to daemon.conf
    - Replaced racist pty jargon in code with more descriptive terms

0.6.4 (20100612)

    - Added more debug statements
    - Fixed typographical errors in documentation
    - Fixed bug: need double-fork on Linux (spotted by Joey Hess joeyh at debian.org)
    - Updated makefiles to work with recent versions of GNU make
    - Updated to avoid new warnings in recent versions of gcc
    - Fixed pidfile race condition (with Hilko Bengen bengen at hilluzination.de)

centos-8-vpnweb.sh
#!/bin/bash -e

sed -i "s/1024/3072 -b 1024 -d 1024 -i 1024/g" /usr/lib/systemd/system/haveged.service
systemctl --quiet daemon-reload && systemctl restart haveged && sudo systemctl --quiet restart sysstat

# Point us at the development environment.
tee --append /etc/hosts <<-EOF > /dev/null
192.168.221.246 api.centos.local
192.168.221.242 vpn.centos.local
192.168.221.242 242.vpn.centos.local
192.168.221.243 243.vpn.centos.local
192.168.221.244 244.vpn.centos.local
192.168.221.245 245.vpn.centos.local
EOF

dnf -q -y module enable go-toolset 1>/dev/null
dnf -q -y install jq git gcc curl make expect golang coreutils gnutls-utils python3-jinja2 python3-netaddr python3-yaml python3-six 1>/dev/null

[ -d $HOME/daemon ] && rm --force --recursive $HOME/daemon
cd $HOME && git clone --quiet https://github.com/lavabit/pahoehoe.git $HOME/daemon && cd $HOME/daemon && FILTER_BRANCH_SQUELCH_WARNING=1 git filter-branch --subdirectory-filter daemon
./configure && make install && cd $HOME

[ -d $HOME/vpnweb ] && rm --force --recursive $HOME/vpnweb
cd $HOME && git clone --quiet https://github.com/lavabit/pahoehoe.git $HOME/vpnweb && cd $HOME/vpnweb && FILTER_BRANCH_SQUELCH_WARNING=1 git filter-branch --subdirectory-filter vpnweb

[ -d /etc/vpnweb/ ] && rm --force --recursive /etc/vpnweb/
mkdir --parents /etc/vpnweb/public/
mkdir --parents /etc/vpnweb/public/1/
mkdir --parents /etc/vpnweb/public/3/

cat <<-EOF > /etc/default/vpnweb
VPNWEB_AUTH=anon
VPNWEB_PORT=443
VPNWEB_ADDRESS=192.168.221.246
VPNWEB_REDIRECT=https://lavabit.com/
VPNWEB_AUTH_SECRET="`mkpasswd -l 32 -d 6 -C 6 -s 0 -d 4`"
VPNWEB_API_PATH=/etc/vpnweb/public/
VPNWEB_PROVIDER_CA=/etc/vpnweb/ca-cert.pem
VPNWEB_TLSCRT=/etc/vpnweb/tls-cert.pem
VPNWEB_TLSKEY=/etc/vpnweb/tls-key.pem
VPNWEB_CACRT=/etc/vpnweb/ca-cert.pem
VPNWEB_CAKEY=/etc/vpnweb/ca-key.pem
EOF

cat <<-EOF > /etc/vpnweb/ca.cfg
organization = "Lavabit LLC"
unit = "Lavabit Developer Proxy"
state = "Texas"
country = US
email = "support@centos.local"
cn = "centos.local"
dc = "centos.local"
serial = 001
activation_date = "2021-01-01 10:00:00"
expiration_date = "2031-12-31 23:59:59"
dns_name = "centos.local"
ca
signing_key
encryption_key
cert_signing_key
EOF

cat <<-EOF > /etc/vpnweb/tls-cert.cfg
organization = "Lavabit LLC"
unit = "Lavabit Developer Proxy"
state = "Texas"
country = US
email = "support@centos.local"
cn = "api.centos.local"
dc = "centos.local"
serial = 002
activation_date = "2021-01-01 12:00:00"
expiration_date = "2031-12-31 12:00:00"
dns_name = "api.centos.local"
signing_key
encryption_key
tls_www_server
EOF

# CA
certtool --stdout-info --sec-param=ultra --generate-privkey --outfile /etc/vpnweb/ca-key.pem 1>/dev/null
certtool --stdout-info --sec-param=ultra --generate-self-signed --load-privkey /etc/vpnweb/ca-key.pem --outfile /etc/vpnweb/ca-cert.pem --template /etc/vpnweb/ca.cfg 1>/dev/null

# TLS
certtool --stdout-info --sec-param=ultra --generate-privkey --outfile /etc/vpnweb/tls-key.pem --template /etc/vpnweb/tls-cert.cfg 1>/dev/null
certtool --stdout-info --sec-param=ultra --generate-request --load-privkey /etc/vpnweb/tls-key.pem --outfile /etc/vpnweb/tls-request.pem --template /etc/vpnweb/tls-cert.cfg 1>/dev/null
certtool --stdout-info --sec-param=ultra --generate-certificate --load-request=/etc/vpnweb/tls-request.pem --load-privkey /etc/vpnweb/tls-key.pem --outfile /etc/vpnweb/tls-cert.pem \
  --load-ca-certificate /etc/vpnweb/ca-cert.pem --load-ca-privkey /etc/vpnweb/ca-key.pem --template /etc/vpnweb/tls-cert.cfg 1>/dev/null

rm --force /etc/vpnweb/ca.cfg
rm --force /etc/vpnweb/tls-cert.cfg
rm --force /etc/vpnweb/tls-request.pem

export FINGERPRINT="`certtool --stdout-info --certificate-info --infile /etc/vpnweb/ca-cert.pem | grep 'sha256:' | head -1 | tr -d '[:space:]' | awk -F':' '{print $2}'`"

cat <<-EOF > /etc/vpnweb/provider.yaml
auth: anon

openvpn:
    - auth: SHA256
    - cipher: AES-256-GCM
    - keepalive: "10 30"
    - tls-cipher: TLS-ECDHE-RSA-WITH-AES-256-GCM-SHA384
    - tun-ipv6: true
    - dev: tun
    - sndbuf: 0
    - rcvbuf: 0
    - nobind: true
    - persist-key: true
    - key-direction: 1

locations:
    - Dallas:
        - name: Dallas
        - country_code: US
        - hemisphere: N
        - timezone: -6

gateways:
    - 242.vpn.centos.local:
        - host: 242.vpn.centos.local
        - ip_address: 192.168.221.242
        - location: Dallas
        - transports:
            - [ "openvpn", "tcp", "443"]
    - 243.vpn.centos.local:
        - host: 243.vpn.centos.local
        - ip_address: 192.168.221.243
        - location: Dallas
        - transports:
            - [ "openvpn", "tcp", "443"]
    - 244.vpn.centos.local:
        - host: 244.vpn.centos.local
        - ip_address: 192.168.221.244
        - location: Dallas
        - transports:
            - [ "openvpn", "tcp", "443"]
    - 245.vpn.centos.local:
        - host: 245.vpn.centos.local
        - ip_address: 192.168.221.245
        - location: Dallas
        - transports:
            - [ "openvpn", "tcp", "443"]

provider:
    - name: "CentOS Test Proxy"
    - description: "CentOS Test Proxy"
    - domain: centos.local
    - api_uri: api.centos.local
    - fingerprint: "SHA256: $FINGERPRINT"
EOF

cat <<-EOF > /etc/vpnweb/public/geoip.json
{
  "ip": "192.168.221.246",
  "cc": "US",
  "city": "Dallas",
  "lat": 32.776,
  "lon": -97.822,
  "gateways": [
    "242.vpn.centos.local",
    "243.vpn.centos.local",
    "244.vpn.centos.local",
    "245.vpn.centos.local"
  ]
}
EOF

cat <<-EOF > /etc/vpnweb/public/1/service.json
{
  "name": "anonymous",
  "description": "anonymous access to the VPN",
  "eip_rate_limit": false
}
EOF

cat <<-EOF > /etc/vpnweb/public/1/configs.json
{
  "services": {
    "eip": "/1/configs/eip-service.json"
  }
}
EOF

cat <<-EOF > /etc/vpnweb/public/3/service.json
{
  "name": "anonymous",
  "description": "anonymous access to the VPN",
  "eip_rate_limit": false
}
EOF

cat <<-EOF > /etc/vpnweb/public/3/configs.json
{
  "services": {
    "eip": "/3/configs/eip-service.json"
  }
}
EOF

GOPATH=$HOME/vpnweb/.go/ go build -a

# $HOME/vpnweb/scripts/gen-shapeshifter-state.py /etc/vpnweb/shapeshifter-state/
# python3 $HOME/vpnweb/scripts/simplevpn.py --file=eip --config=/etc/vpnweb/provider.yaml --template=$HOME/vpnweb/scripts/templates/eip-service.1.json.jinja --obfs4_state=/etc/vpnweb/shapeshifter-state/ | jq -M '.' > /etc/vpnweb/public/1/eip-service.json || echo "ERROR: see /etc/vpnweb/public/1/eip-service.json for output"
# python3 $HOME/vpnweb/scripts/simplevpn.py --file=eip --config=/etc/vpnweb/provider.yaml --template=$HOME/vpnweb/scripts/templates/eip-service.3.json.jinja --obfs4_state=/etc/vpnweb/shapeshifter-state/ | jq -M '.' > /etc/vpnweb/public/3/eip-service.json || echo "ERROR: see /etc/vpnweb/public/3/eip-service.json for output"
python3 $HOME/vpnweb/scripts/simplevpn.py --file=eip --config=/etc/vpnweb/provider.yaml --template=$HOME/vpnweb/scripts/templates/eip-service.1.json.jinja | jq -M '.' > /etc/vpnweb/public/1/eip-service.json || echo "ERROR: see /etc/vpnweb/public/1/eip-service.json for output"
python3 $HOME/vpnweb/scripts/simplevpn.py --file=eip --config=/etc/vpnweb/provider.yaml --template=$HOME/vpnweb/scripts/templates/eip-service.3.json.jinja | jq -M '.' > /etc/vpnweb/public/3/eip-service.json || echo "ERROR: see /etc/vpnweb/public/3/eip-service.json for output"
python3 $HOME/vpnweb/scripts/simplevpn.py --file=provider --config=/etc/vpnweb/provider.yaml --template=$HOME/vpnweb/scripts/templates/provider.json.jinja | jq -M '.' > /etc/vpnweb/public/provider.json || echo "ERROR: see /etc/vpnweb/provider.json for output"

# Kill any existing vpnweb processes and install the new binary.
[ -d /run/vpnweb/ ] && [ -f /run/vpnweb/vpnweb.pid ] && /usr/bin/pkill --pidfile /run/vpnweb/vpnweb.pid

# Create the user, if necessary.
id vpnweb &> /dev/null || adduser --system --shell /sbin/nologin --home-dir /etc/vpnweb/ --no-create-home vpnweb

# Ensure the vpnweb user can access the vpnweb config path.
chown -R vpnweb:vpnweb /etc/vpnweb/

# Setup the pidfile directory.
cat <<-EOF > /usr/lib/tmpfiles.d/vpnweb.conf
D /run/vpnweb 0700 vpnweb vpnweb
EOF

chcon system_u:object_r:lib_t:s0 /usr/lib/tmpfiles.d/vpnweb.conf
[ ! -d /run/vpnweb/ ] && systemd-tmpfiles --create /usr/lib/tmpfiles.d/vpnweb.conf

# Install the binary and the systemd unit file.
install -m 755 $HOME/vpnweb/vpnweb /usr/local/bin/ && setcap 'cap_net_bind_service=+ep' /usr/local/bin/vpnweb

cat <<-EOF > /usr/lib/systemd/system/vpnweb.service
[Unit]
Description=vpnweb
After=network-online.target
Wants=network-online.target

[Service]
Type=simple
User=vpnweb
Group=vpnweb
EnvironmentFile=/etc/default/vpnweb
ExecStart=/usr/local/bin/daemon --inherit --pidfile=/run/vpnweb/vpnweb.pid -- /usr/local/bin/vpnweb -tls=true
ExecStop=bash -c '/usr/bin/test -f /run/vpnweb/vpnweb.pid && /usr/bin/pkill --pidfile /run/vpnweb/vpnweb.pid || echo noop > /dev/null'
ExecReload=bash -c '/usr/bin/test -f /run/vpnweb/vpnweb.pid && /usr/bin/pkill --signal HUP --pidfile /run/vpnweb/vpnweb.pid || echo noop > /dev/null'
LimitNOFILE=65535

# Hardening
ProtectHome=yes
PrivateTmp=yes
CapabilityBoundingSet=CAP_NET_BIND_SERVICE

[Install]
WantedBy=multi-user.target
EOF

chcon system_u:object_r:systemd_unit_file_t:s0 /usr/lib/systemd/system/vpnweb.service

systemctl --quiet daemon-reload
systemctl --quiet enable vpnweb.service && systemctl --quiet start vpnweb.service

echo "VPN web service installed."
echo "Starting unit tests."
sleep 10

curl --fail --silent --show-error --retry 10 --retry-delay 10 --max-time 300 --connect-timeout 300 --output /dev/null --cacert /etc/vpnweb/ca-cert.pem https://api.centos.local/ca.crt || { echo "https://api.centos.local/ca.crt failed ..." ; exit 1 ; }
curl --fail --silent --show-error --retry 10 --retry-delay 10 --max-time 300 --connect-timeout 300 --output /dev/null --cacert /etc/vpnweb/ca-cert.pem https://api.centos.local/provider.json || { echo "https://api.centos.local/provider.json failed ..." ; exit 1 ; }

curl --fail --silent --show-error --retry 10 --retry-delay 10 --max-time 300 --connect-timeout 300 --output /dev/null --cacert /etc/vpnweb/ca-cert.pem https://api.centos.local/1/cert || { echo "https://api.centos.local/1/cert failed ..." ; exit 1 ; }
curl --fail --silent --show-error --retry 10 --retry-delay 10 --max-time 300 --connect-timeout 300 --output /dev/null --cacert /etc/vpnweb/ca-cert.pem https://api.centos.local/1/ca.crt || { echo "https://api.centos.local/1/ca.crt failed ..." ; exit 1 ; }
curl --fail --silent --show-error --retry 10 --retry-delay 10 --max-time 300 --connect-timeout 300 --output /dev/null --cacert /etc/vpnweb/ca-cert.pem https://api.centos.local/1/configs.json || { echo "https://api.centos.local/1/configs.json failed ..." ; exit 1 ; }
curl --fail --silent --show-error --retry 10 --retry-delay 10 --max-time 300 --connect-timeout 300 --output /dev/null --cacert /etc/vpnweb/ca-cert.pem https://api.centos.local/1/service.json || { echo "https://api.centos.local/1/service.json failed ..." ; exit 1 ; }
curl --fail --silent --show-error --retry 10 --retry-delay 10 --max-time 300 --connect-timeout 300 --output /dev/null --cacert /etc/vpnweb/ca-cert.pem https://api.centos.local/1/config/eip-service.json || { echo "https://api.centos.local/1/config/eip-service.json failed ..." ; exit 1 ; }

curl --fail --silent --show-error --retry 10 --retry-delay 10 --max-time 300 --connect-timeout 300 --output /dev/null --cacert /etc/vpnweb/ca-cert.pem https://api.centos.local/3/cert || { echo "https://api.centos.local/3/cert failed ..." ; exit 1 ; }
curl --fail --silent --show-error --retry 10 --retry-delay 10 --max-time 300 --connect-timeout 300 --output /dev/null --cacert /etc/vpnweb/ca-cert.pem https://api.centos.local/3/ca.crt || { echo "https://api.centos.local/3/ca.crt failed ..." ; exit 1 ; }
curl --fail --silent --show-error --retry 10 --retry-delay 10 --max-time 300 --connect-timeout 300 --output /dev/null --cacert /etc/vpnweb/ca-cert.pem https://api.centos.local/3/configs.json || { echo "https://api.centos.local/3/configs.json failed ..." ; exit 1 ; }
curl --fail --silent --show-error --retry 10 --retry-delay 10 --max-time 300 --connect-timeout 300 --output /dev/null --cacert /etc/vpnweb/ca-cert.pem https://api.centos.local/3/service.json || { echo "https://api.centos.local/3/service.json failed ..." ; exit 1 ; }
curl --fail --silent --show-error --retry 10 --retry-delay 10 --max-time 300 --connect-timeout 300 --output /dev/null --cacert /etc/vpnweb/ca-cert.pem https://api.centos.local/3/config/eip-service.json || { echo "https://api.centos.local/3/config/eip-service.json failed ..." ; exit 1 ; }

curl --fail --silent --show-error --insecure --location --retry 10 --retry-delay 10 --max-time 300 --connect-timeout 300 --output /dev/null --cacert /etc/vpnweb/ca-cert.pem https://api.centos.local/ || { echo "https://api.centos.local/ redirection failed ..." ; exit 1 ; }
curl --fail --silent --show-error --insecure --location --retry 10 --retry-delay 10 --max-time 300 --connect-timeout 300 --output /dev/null --cacert /etc/vpnweb/ca-cert.pem https://api.centos.local/$RANDOM || { echo "https://api.centos.local/RANDOM redirection failed ..." ; exit 1 ; }

for i in {1..100}; do
  URL="`cat /dev/urandom | tr -dc '[:alnum:]\#\\\/\.\-\_' | fold -w 32 | head -n 1`"
  curl --fail --silent --show-error --globoff --insecure --location --retry 10 --retry-delay 10 --max-time 300 --connect-timeout 300 --output /dev/null --cacert /etc/vpnweb/ca-cert.pem "https://api.centos.local/${URL}" || { echo "https://api.centos.local/${URL} redirection failed" ; exit 1 ; }
done

echo "Unit tests completed."

# Cleanup.
rm --force --recursive $HOME/go/ $HOME/.cache/ $HOME/vpnweb/ $HOME/daemon/
dnf -q -y module disable go-toolset && dnf -q -y remove jq git gcc make golang expect gnutls-utils python3-jinja2 python3-netaddr python3-yaml patch 1>/dev/null
